code/game/mecha/mech_fabricator.dm

```diff
@@ -281,151 +320,103 @@
-			if("parts")
-				left_part += output_parts_list(part_set)
-				left_part += "<hr><a href='?src=[UID()];screen=main'>Return</a>"
-	dat = {"
-
-			  <title>[name]</title>
-				<style>
-				.res_name {font-weight: bold; text-transform: capitalize;}
-				.red {color: #f00;}
-				.part {margin-bottom: 10px;}
-				.arrow {text-decoration: none; font-size: 10px;}
-				body, table {height: 100%;}
-				td {vertical-align: top; padding: 5px;}
-				html, body {padding: 0px; margin: 0px;}
-				h1 {font-size: 18px; margin: 5px 0px;}
-				</style>
-				<script language='javascript' type='text/javascript'>
-				[js_byjax]
-				</script>
-
-				<table style='width: 100%;'>
-				<tr>
-				<td style='width: 65%; padding-right: 10px;'>
-				[left_part]
-				</td>
-				<td style='width: 35%; background: #000;' id='queue'>
-				[list_queue()]
-				</td>
-				<tr>
-				</table>"}
-	var/datum/browser/popup = new(user, "mecha_fabricator", name, 1000, 600)
-	popup.set_content(dat)
-	popup.open(0)
-	onclose(user, "mecha_fabricator")
-	return
-
 /obj/machinery/mecha_part_fabricator/Topic(href, href_list)
 	if(..())
 		return 1
-	var/datum/topic_input/afilter = new /datum/topic_input(href,href_list)
-	if(href_list["part_set"])
-		var/tpart_set = afilter.getStr("part_set")
-		if(tpart_set)
-			if(tpart_set=="clear")
-				part_set = null
-			else
-				part_set = tpart_set
-				screen = "parts"
-	if(href_list["part"])
-		var/T = afilter.getStr("part")
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
+
+	if(href_list["menu"]) // Main menu
+		screen = MECHFAB_MAIN_MENU
+
+	if(href_list["category"]) // Inside a "category": cyborg parts, medical etc.
+		selected_category = href_list["category"]
+		screen = MECHFAB_CATEGORY_MENU
+
+	if(href_list["search"])
+		if(href_list["to_search"])
+			temp_search = href_list["to_search"]
+		if(!temp_search)
+			return
+		matching_designs.Cut()
+
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
 			if(D.build_type & fabricator_type)
-				if(D.id == T)
-					if(!processing_queue)
-						build_part(D)
-					else
-						add_to_queue(D)
-					break
+				if(findtext(D.name, temp_search))
+					matching_designs.Add(D)
+		screen = MECHFAB_SEARCH_MENU
+
 	if(href_list["add_to_queue"])
-		var/T = afilter.getStr("add_to_queue")
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
+		var/design = href_list["add_to_queue"]
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
 			if(D.build_type & fabricator_type)
-				if(D.id == T)
+				if(D.id == design)
 					add_to_queue(D)
 					break
-		return update_queue_on_page()
+
+	if(href_list["category_add_all"])
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
+			if(D.build_type & fabricator_type && selected_category in D.category)

```

Comment:
Can you pass `selected_category` from the UI here?
Situation: two people operating the same fabricator, one select category A, second switches to category B, the UI does not update for person 1 for whatever reason, person 1 clicks "add all", and adds everything from cat B, which is surprising.

Reply:
After doing some testing, I don't think this will ever happen. Even if the UI doesn't update for the second person, as soon as he clicks to go to a new category, `selected_category` gets updated, which means the adding would fail for the other person (if they are still somehow are viewing the old category).

However, I can pass `data.selected_category` from nano for extra safety, but I'm not sure its necessary.